cpp/dynamic_community_detection_module/algorithm/dynamic_community_detection.cpp

```diff
@@ -0,0 +1,370 @@
+  double sum_Ps = 0;
+
+  for (const auto [label, _] : node_label_Ps) {
+    double inflated_node_label_Ps = pow(node_label_Ps[label], exponent);
+    node_label_Ps[label] = inflated_node_label_Ps;
+    sum_Ps += inflated_node_label_Ps;
+  }
+
+  for (const auto [label, _] : node_label_Ps) {
+    node_label_Ps[label] /= sum_Ps;
+  }
+}
+
+void LabelRankT::Cutoff(
+    std::unordered_map<std::uint64_t, double>& node_label_Ps,
+    double min_value) {
+  std::vector<std::uint64_t> to_be_removed;
+
+  for (const auto [label, P] : node_label_Ps) {
+    if (P < min_value) to_be_removed.push_back(label);
+  }
+
+  for (const auto label : to_be_removed) {
+    node_label_Ps.erase(label);
+  }
+}
+
+std::pair<bool, std::uint64_t> LabelRankT::Iteration(
+    bool incremental, std::unordered_set<std::uint64_t> changed_nodes,
+    std::unordered_set<std::uint64_t> to_delete) {
+  std::unordered_map<std::uint64_t, std::unordered_map<std::uint64_t, double>>
+      updated_label_Ps;
+
+  bool none_updated = true;
+  std::uint64_t most_updates = 0;
+
+  for (const auto node_id : NodesMemgraphIDs()) {
+    if (incremental) {
+      bool was_updated = changed_nodes.count(node_id) > 0 ? true : false;
+      bool was_deleted = to_delete.count(node_id) > 0 ? true : false;
+
+      if (!was_updated || was_deleted) continue;
+    }
+
+    // node selection (a.k.a. conditional update)
+    if (!DistinctEnough(node_id, similarity_threshold)) continue;
+    none_updated = false;
+
+    // label propagation
+    std::unordered_map<std::uint64_t, double> updated_node_label_Ps =
+        Propagate(node_id);
+
+    // inflation
+    Inflate(updated_node_label_Ps, exponent);
+
+    // cutoff
+    Cutoff(updated_node_label_Ps, min_value);
+
+    updated_label_Ps.insert({node_id, updated_node_label_Ps});
+  }
+
+  for (const auto [node, updated_node_label_Ps] : updated_label_Ps) {
+    label_Ps[node] = updated_node_label_Ps;
+    times_updated[node]++;
+
+    if (times_updated[node] > most_updates) most_updates = times_updated[node];
+  }
+
+  return {none_updated, most_updates};
+}
+/* #endregion */
+
+std::unordered_map<std::uint64_t, std::int64_t> LabelRankT::CalculateLabels(
+    std::unique_ptr<mg_graph::Graph<>>& graph,
+    std::unordered_set<std::uint64_t> changed_nodes,
+    std::unordered_set<std::uint64_t> to_delete, bool persist) {
+  this->graph = std::move(graph);
+
+  bool incremental = changed_nodes.size() >= 1 ? true : false;
+  if (incremental) {
+    RemoveDeletedNodes(to_delete);
+    for (const auto node_id : changed_nodes) {
+      SetStructures(node_id);
+    }
+  } else {
+    label_Ps.clear();
+    sum_w.clear();
+    times_updated.clear();
+    for (const auto node_id : NodesMemgraphIDs()) {
+      SetStructures(node_id);
+    }
+  }
+
+  for (std::uint64_t i = 0; i < this->max_iterations; i++) {
+    auto [none_updated, most_updates] =
+        Iteration(incremental, changed_nodes, to_delete);
```

Comment:
`NodesMemgraphIDs` is called in `Iteration`. Because it is a quite expensive operation, it might worth to precalculate it before `Iteration` is called and pass it as a parameter.

Reply:
Thanks for this! I’ve now avoided calling `NodesMemgraphIDs` more than once per algorithm execution.